kythe/cxx/common/kzip_writer.cc

```diff
@@ -78,31 +69,53 @@ absl::string_view Basename(absl::string_view path) {
 }  // namespace
 
 /* static */
-StatusOr<IndexWriter> KzipWriter::Create(absl::string_view path) {
+StatusOr<IndexWriter> KzipWriter::Create(absl::string_view path,
+                                         KzipEncoding encoding) {
   int error;
   if (auto archive =
           zip_open(std::string(path).c_str(), ZIP_CREATE | ZIP_EXCL, &error)) {
-    return IndexWriter(absl::WrapUnique(new KzipWriter(archive)));
+    return IndexWriter(absl::WrapUnique(new KzipWriter(archive, encoding)));
   }
   return libzip::Error(error).ToStatus();
 }
 
 /* static */
 StatusOr<IndexWriter> KzipWriter::FromSource(zip_source_t* source,
+                                             KzipEncoding encoding,
                                              const int flags) {
   libzip::Error error;
   if (auto archive = zip_open_from_source(source, flags, error.get())) {
-    return IndexWriter(absl::WrapUnique(new KzipWriter(archive)));
+    return IndexWriter(absl::WrapUnique(new KzipWriter(archive, encoding)));
   }
   return error.ToStatus();
 }
 
-KzipWriter::KzipWriter(zip_t* archive) : archive_(archive) {}
+KzipWriter::KzipWriter(zip_t* archive, KzipEncoding encoding)
+    : archive_(archive), encoding_(encoding) {}
 
 KzipWriter::~KzipWriter() {
   DCHECK(archive_ == nullptr) << "Disposing of open KzipWriter!";
 }
 
+// Creates entries for the three directories if not already present.
+Status KzipWriter::InitializeArchive(zip_t* archive) {
+  std::vector<std::string> dirs = {kRoot, kFileRoot};
+  if (encoding_ == KzipEncoding::Json || encoding_ == KzipEncoding::All) {
+    dirs.emplace_back(kJsonUnitRoot);
```

Comment:
minor: prefer push_back to emplace_back when both work.  The latter will also call explicit constructors so should generally be reserved where the additional power is required.